src/Control/Parallel/Class.purs

```diff
@@ -0,0 +1,178 @@
+  stall :: forall a. m a
+  race :: forall a. m a -> m a -> m a
+
+unsafeWithRef :: forall eff a. Eff (ref :: REF | eff) a -> Eff eff a
+unsafeWithRef = unsafeInterleaveEff
+
+instance monadParContT :: MonadPar (ContT Unit (Eff eff)) where
+  par f ca cb = ContT \k -> do
+    ra <- unsafeWithRef (newRef Nothing)
+    rb <- unsafeWithRef (newRef Nothing)
+
+    runContT ca \a -> do
+      mb <- unsafeWithRef (readRef rb)
+      case mb of
+        Nothing -> unsafeWithRef (writeRef ra (Just a))
+        Just b -> k (f a b)
+
+    runContT cb \b -> do
+      ma <- unsafeWithRef (readRef ra)
+      case ma of
+        Nothing -> unsafeWithRef (writeRef rb (Just b))
+        Just a -> k (f a b)
+
+instance monadParExceptT :: MonadPar m => MonadPar (ExceptT e m) where
+  par f (ExceptT e1) (ExceptT e2) = ExceptT (par (lift2 f) e1 e2)
+
+instance monadParReaderT :: MonadPar m => MonadPar (ReaderT e m) where
+  par f (ReaderT r1) (ReaderT r2) = ReaderT \r -> par f (r1 r) (r2 r)
+
+instance monadRaceContT :: MonadRace (ContT Unit (Eff eff)) where
+  stall = ContT \_ -> pure unit
+  race c1 c2 = ContT \k -> do
+    done <- unsafeWithRef (newRef false)
+
+    runContT c1 \a -> do
+      b <- unsafeWithRef (readRef done)
+      if b
+        then pure unit
+        else do
+          unsafeWithRef (writeRef done true)
+          k a
+
+    runContT c2 \a -> do
+      b <- unsafeWithRef (readRef done)
+      if b
+        then pure unit
+        else do
+          unsafeWithRef (writeRef done true)
+          k a
+
+instance monadRaceExceptT :: MonadRace m => MonadRace (ExceptT e m) where
+  stall = ExceptT stall
+  race (ExceptT e1) (ExceptT e2) = ExceptT (race e1 e2)
+
+instance monadRaceReaderT :: MonadRace m => MonadRace (ReaderT e m) where
+  stall = ReaderT \_ -> stall
+  race (ReaderT r1) (ReaderT r2) = ReaderT \r -> race (r1 r) (r2 r)
+
+-- | The `Parallel` type constructor provides `Applicative` and `Alternative`
+-- | instances for any type monad with `MonadPar` and `MonadRace` instances
+-- | respectively.
+-- |
+-- | - The definition of `apply` from `Apply` runs two computations in parallel and applies
+-- |   a function when both complete.
+-- | - The definition of `alt` from the `Alt` type class runs two computations in parallel
+-- |   and returns the result of the computation which completes first.
+-- |
+-- | Parallel sections of code can be embedded in sequential code by using
+-- | the `parallel` and `runParallel` functions:
+-- |
+-- | ```purescript
+-- | loadModel :: ContT Unit (Eff (ajax :: AJAX)) Model
+-- | loadModel = do
+-- |   token <- authenticate
+-- |   runParallel $
+-- |     Model <$> parallel (get "/products/popular/" token)
+-- |           <*> parallel (get "/categories/all" token)
+-- | ```
+newtype Parallel m a = Parallel (m a)
+
+-- | Lift a computation to be run in parallel from a computation in the
+-- | corresponding `Monad`.
+parallel :: forall m a. m a -> Parallel m a
+parallel = Parallel
```

Comment:
Should we impose a `MonadPar m` constraint here perhaps? There's no advantage to being able to construct `Parallel` without that constraint, is there? I guess the `Functor` instance would also need tightening to `MonadPar m` for consistency.